content/news/2022-12-03-dev_4.0_branch/index.md

```diff
@@ -0,0 +1,125 @@
+![Proposed API Buffer Performance](buffer_benchmarks.png "Proposed Buffer API operations per second")
+
+
+Getting to this point, though, requires much more information to be provided to the block at compile time through templating.  This may appear cumbersome at first, but much of this code could become part of the yml->c++ code generation step and get right to the work method as quickly as before
+
+
+#### Work implementations on non-standard types
+
+Another benefit of the underlying buffer structure being strongly typed and not just bytes is that it can be used to pass more complex data structures using the circular buffers into the work methods of blocks.  For example, backpressure based PDU flowgraphs could be implemented just by defining a port as type PDU for asynchronous signal processing, and not merely through using the asynchronous message queues.  This will allow a more intelligent and efficient scheduling of packet based flowgraphs.
+
+
+### Less overhead
+
+When passing large amounts of data into a block with a computationally complex work function, the scheduler and lower level computational overhead from things like virtual table lookups are overshadowed by the computation done in the block.  But making the low level API as efficient and close to zero overhead as possible opens the door to different scheduling policies and paradigms that are able to process smaller chunks of data and maintain higher throughput while introducing less latency through the flowgraph.
+
+With modern programming techniques, many of the operations that are performed in calculating the necessary information to perform work(), for example, can be more greatly optimized at compile time through constexpr.  Additionally, removing virtual inheritance from the hot path - namely, access to the circular buffers - can further reduce overhead.
+
+### More intuitive design
+
+One of the concerns about current GNU Radio block design is the sheer number of considerations necessary to ensure the block interacts appropriately with the scheduler.  This can be addressed effectively by enforcing a stronger separation of concerns in the hierarchy of classes and interfaces.  For example, having well defined buffer object associated with the port keeps the management of data in one place and not detached from the block structure. 
+
+Also proposed is a realignment of the entities in the API to be cleaner an more intuitive as seen below:
+
+![Separation of Concerns](separation_of_concerns.png "Separation of Concerns")
+
+Just to give a rough idea of some of the proposed mechanisms, below is a snapshot of what a type safe block and port construct could look like in practice:
+```c++
+template<typename T, typename U = T, … >
+requires (gr::util::is_one_of<T, supported_type>::value)
+class myCopyBlock : public gr::node<"copy", gr::IN<"IN", T>, gr::OUT<"OUT", T>, int32_t, float, std::complex<float>> {
```

Comment:
Should this be `gr::OUT<"OUT", U>` (vs T), or why is the `U = T` there?

Reply:
I'm not sure - this is copied directly from the slides